httputil/request_test.go

```diff
@@ -0,0 +1,76 @@
+package httputil
+
+import (
+	"github.com/stretchr/testify/assert"
+	"net/http"
+	"testing"
+)
+
+func TestGetAutTypeNoAuthHeaders(t *testing.T) {
+	headers := map[string][]string{}
+
+	authType := GetAuthType(httpResponseFromHeaders(headers))
+
+	assert.Equal(t, BasicAuthType, authType)
+}
+
+func TestGetAutTypewwAuthenticateBasicNtlmBearer(t *testing.T) {
+	headers := map[string][]string{
+		"WWW-Authenticate": {"Basic", "NTLM", "Bearer"},
+	}
+
+	authType := GetAuthType(httpResponseFromHeaders(headers))
+
+	assert.Equal(t, NtlmAuthType, authType)
+}
+
+func TestGetAutTypeNoWwwAuthenticateLfsAuthenticateNtlm(t *testing.T) {
+	headers := map[string][]string{
+		"LFS-Authenticate": {"Basic", "NTLM", "Bearer"},
+	}
+
+	authType := GetAuthType(httpResponseFromHeaders(headers))
+
+	assert.Equal(t, NtlmAuthType, authType)
+}
+
+func TestGetAutTypeNoWwwAuthenticateLfsAuthenticateBasicNtlm(t *testing.T) {
+	headers := map[string][]string{
+		"LFS-Authenticate": {"Basic", "Ntlm"},
+	}
+
+	authType := GetAuthType(httpResponseFromHeaders(headers))
+
+	assert.Equal(t, NtlmAuthType, authType)
+}
+
+func TestGetAutTypeWwwAuthenticateBasicNegotiate(t *testing.T) {
+	headers := map[string][]string{
+		"Www-Authenticate": {"Basic", "Negotiate"},
+	}
+
+	authType := GetAuthType(httpResponseFromHeaders(headers))
+
+	assert.Equal(t, NtlmAuthType, authType)
+}
+
+func TestGetAutTypeWwwAuthenticateBasicLfsAuthenticateNtlm(t *testing.T) {
+	headers := map[string][]string{
+		"WWW-Authenticate": {"Basic"},
+		"LFS-Authenticate": {"Ntlm"},
+	}
+
+	authType := GetAuthType(httpResponseFromHeaders(headers))
+
+	assert.Equal(t, NtlmAuthType, authType)
+}
+
+func httpResponseFromHeaders(headers map[string][]string) *http.Response {
```

Comment:
OK, I'm going going to make one more annoying suggestion about this function. I hope you'll forgive me 😜 

All of the tests in this file follow the same structure:

1. Initialize some headers.
2. Get the resulting auth type from those headers.
3. Assert that it equals some other auth type.

I'm proposing the following test-case struct, and the following test structure:

```go
type AuthenticateHeaderTestCase struct {
        AuthType string
        Headers map[string][]string
}
```

func (c *AuthenticateHeaderTestCase) Assert(t *testing.T) {
        t.Log("lfs/httputil: asserting auth type: %q for: %v", c.AuthType, c.Headers)

        assert.Equal(t, c.AuthType, GetAuthType(c.HttpResponse())
}

func (c *AuthenticateHeaderTestCase) HttpResponse() *http.Response {
        res := &http.Response{Header: make(http.Header)}
        for k, vv := range c.Headers {
                for _, v := range vv {
                        res.Header.Add(k, v)
                }
        }

        return res
}
```

and the following test structure:

```go
func TestAuthenticationCases(t *testing.T) {
        for _, c := range []AuthenticateHeaderTestCase{
                {NtlmAuthType, map[string]string{"Www-Authenticate" : {"Basic", "Negotiate"}}},
                {NtlmAuthType, map[string]string{"LFS-Authenticate" : {"Basic", "Ntlm"}}},
                // ...
        } {
                c.Assert(t)
        }
}
```

This is akin to the test-case pattern found in the [`batcher_test.go`](https://github.com/github/git-lfs/blob/d25c0eb03e45d788f971ebd50f4a47772db39ac7/lfs/batcher_test.go) file.

Feel free to not go with this implementation, I don't mind either way :smile:. What you have right now works great, and is cool with me, too :+1: